experimental/PropertyDDS/packages/property-dds/src/propertyTreeExtFactories.ts

```diff
@@ -0,0 +1,52 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+    IChannelAttributes,
+    IFluidDataStoreRuntime,
+    IChannelServices,
+    IChannelFactory,
+} from "@fluidframework/datastore-definitions";
+import { SharedPropertyTreeOptions } from "./propertyTree";
+import { CompressedCommPropertyTree } from "./propertyTreeExt";
+
+export class CompressedCommPropertyTreeFactory implements IChannelFactory {
+    public static readonly Type = "CompressedPropertyTree:84534a0fe613522101f6";
+
+    public static readonly Attributes: IChannelAttributes = {
+        type: CompressedCommPropertyTreeFactory.Type,
+        snapshotFormatVersion: "0.1",
+        packageVersion: "0.0.1",
+    };
+
+    public get type() {
+        return CompressedCommPropertyTreeFactory.Type;
+    }
+
+    public get attributes() {
+        return CompressedCommPropertyTreeFactory.Attributes;
+    }
+
+    public async load(
+        runtime: IFluidDataStoreRuntime,
+        id: string,
+        services: IChannelServices,
+        attributes: IChannelAttributes,
+        url?: string,
+    ): Promise<CompressedCommPropertyTree> {
+        const options = {};
+        const instance = new CompressedCommPropertyTree(id, runtime, attributes, options as SharedPropertyTreeOptions);
+        await instance.load(services);
+        return instance;
+    }
+
+    public create(document: IFluidDataStoreRuntime, id: string, requestUrl?: string): CompressedCommPropertyTree {
+        const options = {};
+        const cell = new CompressedCommPropertyTree(id, document,
+            this.attributes, options as SharedPropertyTreeOptions);
+        cell.initializeLocal();
+        return cell;
+    }
```

Comment:
@ruiterr - I had one idea on how to do this without a subclass:

1.  Change the PropertyTree ctor so that it accepts an IPropertyTreeConfig as an argument.
2.  Have the end user create a custom IChannelFactory instance for their desired configuration that creates/passes their desired IPropertyTreeConfig.

The IPropertyTreeConfig is the extensibility point, which possibly looks something like:
```ts
{
    messageEncoder: { encode: (message) => encoded, decode: (encoded) => message };
    summaryEncoder: { encode: (summary) => encoded, decode: (encoded) => summary };
}
```
Combined with a helper function that generates a factory instance for a given IPropertyTreeConfig, it might make it easier to experiment with a matrix of configurations in the context of a benchmark or test.

The one hazard with encouraging the end user to create their own factory is that it's important that each client has the same mapping from the id in the factory to the same configuration/code.  (That's why the normal usage pattern is to create a singleton factory instance that end users reuse.)

Reply:
To give a bit of background: the motivation for this approach is that it will make it possible to have multiple different flags / options. 

If you implement an optional feature via a sub-class, you cannot combine it with another optional feature (let's say partial checkouts), without creating subclasses for all pair-wise combinations. If instead the optional code is within separate classes, you can mix multiple of those.

Apart from this change, I am fine with having this extensibility built into the PropertyDDS and performing the compression within the class (for the moment). Event though, I agree that it would be nicer to have this on a lower level within Fluid, that would probably be a major refactoring.